### PR TITLE
Binomial distribution

### PR DESCRIPTION
**Pull request by eightysteele**
# Purpose

This PR adds support for the binomial distribution. 

# Log-likelihood

The log PMF for the binomial distribution gives the log probability of observing $k$ successes out of $n$ trials:

$P(X = k) = \binom{n}{k} p^k (1-p)^{n-k}$

For computing that binomial coefficient, I used the existing `log-gamma-fn`. Otherwise for large values of $n$ and $k$ it would have overflown on factorials.

# Testing

In `distribution_test.clj` I tested boundary conditions on $n$, $p$, and $k$ and verified some different properties of the distribution that should definitely hold (sum of probabilities, symmetry when $p=0$, mean and variance consistency). Both `bb test:clj` and `bb test:cljs` are passing for these.

I also spot checked a range of values against SciPy and Gen.jl (within a precision threshold of  `1e-9` of expected values). One of theses tests is currently failing because the difference between expected and actual is right at the threshold boundary. If this is problematic, let me know and I can dig in!

```clojure
      ;; TODO: failing test (off by 1.9e-9)
      ;; expected: (ish? -3.222306954272568 (dist/logpdf (->binomial 1000000 0.0001) 100))
      ;; actual: (not (ish? -3.222306954272568 -3.2223069561241857))
      (is (ish? -3.222306954272568 (dist/logpdf (->binomial 1000000 0.0001) 100)))
``` 
# Next steps

A spot check and some feedback would be great. It's entirely possible that I'm missing something here. Happy to make changes.

### CODE REVIEW

File: src/gen/distribution/java_util.clj [hunk 1]
```diff
@@ -25,6 +25,15 @@
   (d/logpdf [_ v]
     (ll/bernoulli p v)))
 
+(defrecord Binomial [^SplittableRandom rnd n p]
+  d/Sample
+  (sample [_]
+    (.nextBinomial rnd n p))
```
Comment (sritchie) [inline src/gen/distribution/java_util.clj:31]: no way, awesome! Is this a thing? I don't see it in the javadocs... https://docs.oracle.com/javase/8/docs/api/java/util/SplittableRandom.html

File: src/gen/distribution/math/log_likelihood.cljc [hunk 2]
```diff
@@ -97,6 +97,46 @@
...
+  distribution](https://en.wikipedia.org/wiki/Binomial_distribution)
+  parameterized by `n` (number of trials) and `p` (probability of success in
+  each trial) at the value `v` (number of successes)."
+  [n p v]
+  {:pre [(integer? n)
+         (integer? v)
+         (>= v 0)
+         (>= n v)
+         (<= 0 p 1)]}
+  (cond
```
Comment (sritchie) [inline src/gen/distribution/math/log_likelihood.cljc:128]: you might use `case` here to make this more efficient.

File: src/gen/distribution/math/log_likelihood.cljc [hunk 3]
```diff
@@ -97,6 +97,46 @@
   {:pre [(<= 0 p 1)]}
   (Math/log (if v p (- 1.0 p))))
 
+(defn log-fact
+  "Returns the natural logarithm of `x` factorial."
+  [x]
+  {:pre [(>= x 0)]}
+  (log-gamma-fn (inc x)))
+
+(defn log-bico
```
Comment (sritchie) [inline src/gen/distribution/math/log_likelihood.cljc:106]: let's make this private, or add `^:no-doc` to keep it out of the public API

Comment (sritchie) [reply src/gen/distribution/math/log_likelihood.cljc:106]: I would actually make this a local fn with `letfn` instead of making it its own thing. All of these preconditions are going to run every time, if you put it inside `binomial` you can skip running those since you're calling it in a place where you know you pass the preconditions.

File: test/gen/distribution_test.cljc [hunk 4]
```diff
@@ -59,6 +59,87 @@
...
+    (is 0 (dist/logpdf (->binomial 10 0) 0)))
+ 
+  (testing "when p = 0 and v > 0, probability is 0, log(0) = -Inf"
+    (is ##-Inf (dist/logpdf (->binomial 10 0) 1)))
+
+  (testing "when p = 1 and v = n, probability is 1, log(1) = 0"
+    (is 0 (dist/logpdf (->binomial 10 1) 10)))
+
+  (testing "when p = 1 and v < n, probability is 0, log(0) = -Inf"
+    (is ##-Inf(dist/logpdf (->binomial 10 0) 1)))
```
Comment (sritchie) [inline test/gen/distribution_test.cljc:74]: whoops, missing a space

File: test/gen/distribution_test.cljc [hunk 5]
```diff
@@ -59,6 +59,87 @@
...
+
+  ;; properties...
+  (testing "sum of probabilities equals 1"
+    (with-comparator (within 1e-9)
+      (let [n 100
```
Comment (sritchie) [inline test/gen/distribution_test.cljc:79]: how about making this a generative test? See the `checking` style from other tests.

File: test/gen/distribution_test.cljc [hunk 6]
```diff
@@ -59,6 +59,87 @@
...
+        (is (ish? 1.0 sum-probs)))))
+
+  (testing "symmetric when p = 0.5 such that binomial(k) = binomial(n -k)"
+    (let [n 100
+          p 0.5
+          v 10]
+      (is (dist/logpdf (->binomial n p) v)
+          (dist/logpdf (->binomial n p) (- n v)))))
+
+  (testing "mean and variance consistency where mu = n * p and variance = mu(1 - p)"
```
Comment (sritchie) [inline test/gen/distribution_test.cljc:93]: can you say more, even if just in comments below, about what you are testing here?

File: test/gen/distribution_test.cljc [hunk 7]
```diff
@@ -59,6 +59,87 @@
...
+      (is (ish? 0 (dist/logpdf (->binomial 10 0) 0)))
+      (is (ish? 0 (dist/logpdf (->binomial 10 1) 10)))
+      (is (ish? -2.02597397686619 (dist/logpdf (->binomial 100 0.9) 90)))
+      (is (ish? -52.680257828913156 (dist/logpdf (->binomial 500 0.1) 0)))))
+
+  (testing "spot check against gen logpdf (v0.4.6)"
+    (with-comparator (within 1e-9)
+      (is (ish? -7.133546882067904 (dist/logpdf (->binomial 1000000 0.5) 500000)))
+
+      ;; TODO: failing test (off by 1.9e-9)
```
Comment (sritchie) [inline test/gen/distribution_test.cljc:129]: delete these if unnecessary

**Review by sritchie — CHANGES_REQUESTED** [review]
Nice work, just a few comments!

File: src/gen/distribution/math/log_likelihood.cljc (hunk 2 above)
Comment (eightysteele) [reply src/gen/distribution/math/log_likelihood.cljc]: oh nice, done.

File: src/gen/distribution/math/log_likelihood.cljc (hunk 3 above)
Comment (eightysteele) [reply src/gen/distribution/math/log_likelihood.cljc]: totally. i made `log-fact` and `log-bico` local with `letfn`.

File: test/gen/distribution_test.cljc (hunk 4 above)
Comment (eightysteele) [reply test/gen/distribution_test.cljc:82]: fixed. should `bb lint` normally catch that?

File: test/gen/distribution_test.cljc (hunk 5 above)
Comment (eightysteele) [reply test/gen/distribution_test.cljc]: great idea. i changed a bunch of tests over to generative. done.

File: test/gen/distribution_test.cljc (hunk 6 above)
Comment (eightysteele) [reply test/gen/distribution_test.cljc]: yeah, i added comments with more context. the basic idea is to spot check a couple of well known properties of the distribution. if that seems weird though i can leave them out for sure.

File: test/gen/distribution_test.cljc (hunk 7 above)
Comment (eightysteele) [reply test/gen/distribution_test.cljc]: done. i dialed down the precision threshold from `1e-9` to `1e-6` and all of these tests are now passing. for spot checking against other libraries, that seems like a peaceful threshold but feel free to push back!

**Comment by eightysteele** [comment]
@sritchie thanks for the fast feedback! just pushed updates and responded to comments.

i took a look at that score function in webppl to see how it compared to the lanczos approximation approach in `log-gamma-fn`:

https://github.com/probmods/webppl/blob/dev/src/dists/binomial.ad.js#L66-L107

i took a quick stab at porting to clojure so that i could run a comparison (this can almost certainly be optimized):

```clojure
(defn binomial
  [n p v]
  {:pre [(integer? n)
         (integer? v)
         (>= v 0)
         (>= n v)
         (<= 0 p 1)]}
  (letfn [(log-fact
            [x]
            (if (< x 2)
              0
              (reduce + (map #(Math/log %) (range 2 (inc x))))))
          (log-bico
            [n m o]
            (let [x (reduce + (map #(Math/log %) (range (inc o) (inc n))))]
              (- x (log-fact m))))]
    (let [[m o] (if (< v (- n v)) [v (- n v)] [(- n v) v])]
      (case p
        0 (if (= v 0)
            0.0     ;; log(1)
            ##-Inf) ;; log(0))
        1 (if (= v n)
            0.0     ;; log(1)
            ##-Inf) ;; log(0)
        (+ (log-bico n m o)
           (* v (Math/log p))
           (* (- n v) (Math/log (- 1 p))))))))
 ```

this particular implementation turned out to be pretty slow with a bit of precision loss as $n$ and $v$ got large (7 minutes to run `bb test:clj` compared to 16 seconds with `log-gamma-fn`).

here's my hypothesis. the runtime complexity of  my original `binomial` is basically $O(1)$, driven by the lanczos approximation in `log-gamma-fn`. that's because it involves a fixed number of coefficients and iterations that do not vary with the input. this is assuming constant time for arithmetic operations.

for the `binomial` test above, the complexity is $O(n)$ from the linear scaling of log summations within the range $n - o$ and the factorial computation $O(m)$. with the optimization to minimize computation range, the worst-case scenario scales linearly with $n$, which bumps the computational cost for larger inputs.

all to say, it seems reasonable to keep the original implementation for now.

File: src/gen/distribution/java_util.clj (hunk 1 above)
Comment (eightysteele) [reply src/gen/distribution/java_util.clj]: (i also added `binomial-gf-tests` that generatively samples from a distribution, which would have caught `.nextBinomial`)

**Comment by eightysteele** [comment]
purely for my own learning, i worked out a slightly more efficient implementation that memoizes factorials and swaps map/reduce for loop/recur (to factor out intermediate collections). it's roughly 78% faster than that first version, but still a minute slower than the `log-gamma-fn` version. 

```clojure
(def memoized-log-fact
  (memoize
   (fn [x]
     (loop [i x
            acc 0.0]
       (if (< i 2)
         acc
         (recur (dec i) (+ acc (Math/log i))))))))

(defn log-bico
  [n m o]
  (let [numerator (loop [i (inc o)
                         acc 0.0]
                    (if (> i n)
                      acc
                      (recur (inc i) (+ acc (Math/log i)))))]
    (- numerator (memoized-log-fact m))))

(defn binomial
  [n p v]
  {:pre [(integer? n)
         (integer? v)
         (>= v 0)
         (>= n v)
         (<= 0 p 1)]}
  (let [[m o] (if (< v (- n v)) [v (- n v)] [(- n v) v])]
    (case p
      0 (if (= v 0) 0.0 ##-Inf)
      1 (if (= v n) 0.0 ##-Inf)
      (+ (log-bico n m o)
         (* v (Math/log p))
         (* (- n v) (Math/log (- 1 p)))))))
'''

File: src/gen/distribution/math/log_likelihood.cljc [hunk 8]
```diff
@@ -97,6 +66,36 @@
   {:pre [(<= 0 p 1)]}
   (Math/log (if v p (- 1.0 p))))
 
+(defn binomial
+  "Returns the log-likelihood of a [Binomial
+  distribution](https://en.wikipedia.org/wiki/Binomial_distribution)
+  parameterized by `n` (number of trials) and `p` (probability of success in
+  each trial) at the value `v` (number of successes)."
+  [n p v]
+  {:pre [(integer? n)
```
Comment (sritchie) [inline src/gen/distribution/math/log_likelihood.cljc:75]: I think in ClojureScript, `number?` checks if you have a `goog.math/Integer`... is that right? I want to check on how we did this in Emmy, I feel like there is some other predicate we should use. I could be wrong, flying blind here before a meeting..

File: test/gen/distribution_test.cljc (hunk 4 above)
Comment (sritchie) [reply test/gen/distribution_test.cljc:82]: looks like it's still there, I'm not sure, maybe it should?

**Review by sritchie — APPROVED** [review]

**Comment by sritchie** [comment]
Thanks @eightysteele , this looks great!!

File: src/gen/distribution/math/log_likelihood.cljc (hunk 8 above)
Comment (eightysteele) [reply src/gen/distribution/math/log_likelihood.cljc:75]: yeah not sure, but it doesn't look like cljs checks for `goog.math/Integer`:

https://github.com/clojure/clojurescript/blob/acbefb9b1e79b659b639919fbf96cb3726719e25/src/main/cljs/cljs/core.cljs#L2328-L2334

in emmy i don't see an `integer?` defined, but the value namespace does have predicates like `real?` and `number?` that do use `goog.math/...`:

https://github.com/mentat-collective/emmy/blob/main/src/emmy/value.cljc 

BUT if i understand the compatibility layer between clj and cljs, if we use integer? in clojure code and then compile it to cljs, the cljs version of integer? should be used when the code runs in a js environment. if that's the case, then it's plausible that this is the right predicate to use?

**Comment by eightysteele** [comment]
thanks @sritchie! \o/
